# Integer, Float, and Boolean Data types

This chapter will go deeper into the different integer, float, and boolean data types that are available to the pandas Series and DataFrame. As a reminder, all values in a Series are exactly one data type. Similarly, all values in each column of a DataFrame are exactly one data type.

While the words 'boolean', 'integer', and 'float' are useful descriptions, they are not the technical names of the actual data types. This chapter provides a complete picture of all the exact integer, float, and boolean data types that are available and what they mean. We will be making heavy use of the `astype` method to change data types.

## Constructing a Series

Before getting started with our data type discussion, we will learn how to create a Series using its constructor. All of our previous Series have been created by selecting a single column from a DataFrame. It is also possible to create a Series manually using `pd.Series` and passing it a list of values. Let's create a simple integer Series with a few values.

In [1]:
import pandas as pd
s_int = pd.Series([50, 99, 130])
s_int

0     50
1     99
2    130
dtype: int64

Creating a Series in this manner is often referred to as using the **constructor**, which is a generic programming term describing the creation and instantiation of a new object of a particular type. We used the Series constructor to create a new Series of integers.

## Integer data type

The visual output of a Series shows the data type of the values below it. In this case, it is `'int64'`, which formally represents a 64-bit integer. This data type comes directly from numpy, which allows integers to be either 8, 16, 32, or 64 bits in size. A 64-bit integer can contain up to 2 raised to the 64th power number of integers. Let's see how many numbers this is.

In [2]:
2 ** 64

18446744073709551616

The `int64`  data type has both positive and negative integers. Let's divide the above number by 2 to get the maximum integer allowed.

In [3]:
2 ** 64 // 2

9223372036854775808

numpy has a function called `iinfo` that returns the exact integer information for each integer data type. Pass it the data type as a string to get the information. Note that the range of integers is exactly 2 raised to the 64th power when accounting for 0.

In [4]:
import numpy as np
np.iinfo('int64')

iinfo(min=-9223372036854775808, max=9223372036854775807, dtype=int64)

In [8]:
np.iinfo('int8')

iinfo(min=-128, max=127, dtype=int8)

The span of numbers for 'int8' goes from -128 to 127 or 256 total numbers. This is equivalent to 2 raised to the 8th power.

## Changing data types with `astype`

You can change the data type of a Series with the `astype` method by passing it the string name of the data type. The string name is going to be the base data type, which is 'int' in this case, appended directly to the number of bits (8, 16, 32, or 64).

Below, we change the data type to 'int8'. Notice that the third value now displays as -126 and not its original value of 130. The maximum 8-bit integer is 127, making 130 greater than it by 3. numpy assumes you know what you are doing and does not check that this number goes beyond its maximum. Instead, the number is represented by the third integer greater than the minimum -128 which is -126.

In [10]:
s_int

0     50
1     99
2    130
dtype: int64

In [11]:
s_int.astype('int8')

0     50
1     99
2   -126
dtype: int8

### Default integer types

Unfortunately, the default data type is dependent on the platform that you are using. numpy is built with the C programming language, and uses the size of the C **long** type as its default. For 32-bit Linux, macOS, and Windows machines, this will be 32 bits. For 64-bit Linux and macOS machines, it will be 64 bits. For 64-bit Windows machines this will be 32-bits. The [first two rows of this table][1] show the size of C long types for each platform.

### Windows machines

If you have a Windows machine, you may have noticed that your data type still says 'int64' even though the previous section just said otherwise. This is because we constructed our Series with a list and not a numpy array. Let's construct a numpy array of integers using its constructor, `array`.

[1]: https://en.wikipedia.org/wiki/Integer_(computer_science)#Long_integer

In [12]:
a = np.array([1, 5])
a

array([1, 5])

numpy arrays use the exact same `dtype` attribute as pandas Series objects to access the data type. Windows users should see 'int32' as the data type, while 64-bit Linux and macOS users should see 'int64'.

In [13]:
a.dtype

dtype('int64')

Constructing the Series with a numpy array will show the same data type as above.

In [14]:
pd.Series(a)

0    1
1    5
dtype: int64

## Unsigned Integers

The default integer data types split half of their range between negative and positive integers. It is possible to limit your integers to just the non-negative integers by using the unsigned integer data type abbreviated with the string 'uint'. The same sizes 8, 16, 32, and 64 bits are available. Let's convert the original `s_int` Series to 'uint8'. 

Calling the `astype` method returns a copy of the data, and does not work in-place. We first verify that the `s_int` Series still has data type of 'int64'.

In [15]:
s_int.dtype

dtype('int64')

We now convert the data type to an 8-bit unsigned integer.

In [21]:
s_int.astype('uint8').info(memory_usage='deep')

<class 'pandas.core.series.Series'>
RangeIndex: 3 entries, 0 to 2
Series name: None
Non-Null Count  Dtype
--------------  -----
3 non-null      uint8
dtypes: uint8(1)
memory usage: 135.0 bytes


The last value remains correctly as 130 as the range of this new data type is from 0 to 255. Let's verify this with the `iinfo` method.

In [20]:
s_int.info(memory_usage='deep')

<class 'pandas.core.series.Series'>
RangeIndex: 3 entries, 0 to 2
Series name: None
Non-Null Count  Dtype
--------------  -----
3 non-null      int64
dtypes: int64(1)
memory usage: 156.0 bytes


In [19]:
np.iinfo('uint8')

iinfo(min=0, max=255, dtype=uint8)

It is rare to use unsigned integers, but they are available, and can be useful in specific situations where you would like to save memory. Typically, it won't be necessary to use them, so sticking with the default integer data type should work.

## Nullable integer data type

With the release of pandas version 0.24 in late 2019, a new nullable integer data type became available to pandas users. This new data type allows for missing values to be present in a column of integers. This is a completely separate data type than the normal integer data types. The original integer data types still exist and are unable to contain missing values. Let's verify this by attempting to construct an integer Series with missing values. Notice, that we use the `dtype` parameter to attempt to force the data type to be 'int64'.

In [22]:
pd.Series([50, 99, 130, np.nan], dtype='int64')

ValueError: cannot convert float NaN to integer

If you do not use the `dtype` parameter, then the Series will be constructed, but use the more flexible float64 data type which does allow missing values.

In [23]:
pd.Series([50, 99, 130, np.nan])

0     50.0
1     99.0
2    130.0
3      NaN
dtype: float64

### Constructing a Series of nullable integers

The nullable integer data type is represented by the string 'Int' (as opposed to 'int'). The important distinction is the capitalized first letter I. The same four bit sizes, 8, 16, 32, and 64 are available. Let's construct a Series of nullable integers using the string 'Int64'. 

In [68]:
s_nullable_int = pd.Series([50, 99, 130, np.nan], dtype='Int64')
s_nullable_int

0      50
1      99
2     130
3    <NA>
dtype: Int64

The missing value is visually represented by `<NA>` which is different than `NaN` when the Series was constructed with `float64`. pandas introduced its own missing value object `NA` that is distinct from numpy's `NaN`. pandas will convert any missing value in nullable integer Series to its own `NA` object. Let's use pandas `NA` object directly in the Series construction to show that the same Series is created.

In [69]:
pd.Series([50, 99, 130, pd.NA], dtype='Int64')

0      50
1      99
2     130
3    <NA>
dtype: Int64

### Nullable integers don't exist in numpy

The nullable integer data type is only available for pandas objects. They do not exist in numpy.

### Different behavior of nullable integer

The nullable integer data type exhibits different behavior than the normal integer data type. If attempting to construct a Series with values that are not within its range, an exception will be raised and not wrap around like it did above with numpy data types. This is probably better behavior to prevent mistakes.

In [30]:
pd.Series([50, 99, 130], dtype='Int8')

TypeError: cannot safely cast non-equivalent int64 to int8

Another difference is when using the comparison operators. They evaluate as missing when comparing against a missing value. Here, we compare each value in the nullable integer Series with 100. The last value was originally missing and remains missing after the operation.

In [31]:
nullable_filt = s_nullable_int > 100
nullable_filt

0    False
1    False
2     True
3     <NA>
dtype: boolean

Let's convert this Series so its data type is a 64-bit float.

In [32]:
s_float = s_nullable_int.astype('float64')
s_float

0     50.0
1     99.0
2    130.0
3      NaN
dtype: float64

Notice that the missing value is now a numpy `NaN` and not a pandas `NA`. Executing the same comparison operation yields a Series where the last value is no longer missing. Comparisons with the numpy float data types against a missing value always evaluate as `False`.

In [33]:
filt = s_float > 100
filt

0    False
1    False
2     True
3    False
dtype: bool

Take a look above at the `nullable_filt` Series we created. Notice that the data type is `'boolean'` as opposed to `'bool'` in the `filt` Series. This is a separate new data type, nullable boolean, introduced in pandas 1.0. It will be discussed in detail further below.

### Boolean selection with nullable booleans

Boolean selection works similarly with the nullable boolean data type as it does with with the old `'bool'` data type. Let's use `nullable_filt` to filter the `s_nullable_int` Series. Even though the last value of `nullable_filt` is missing, the selection still works and the missing value is treated as `False`. The same behavior is exhibited with the `query` method.

In [34]:
s_nullable_int[nullable_filt]

2    130
dtype: Int64

### Missing value evaluation with comparison operators

All comparisons with numpy `nan` result in False, while all comparisons with the new pandas `NA` evaluate as `NA`. A few examples involving different comparison operators are presented below for each of the missing value objects.

In [35]:
np.nan > 5, np.nan < 5, np.nan == 5

(False, False, False)

In [36]:
pd.NA > 5, pd.NA < 5, pd.NA == 5

(<NA>, <NA>, <NA>)

### Unsigned nullable integers

Unsigned nullable integers are also available with the string `'UInt'` (capital U and I). Our Series construction that failed with the signed 8-bit nullables from above now works with the unsigned 8-bit nullable integer as 130 is part of the range.

In [37]:
pd.Series([50, 99, 130, pd.NA], dtype='UInt8')

0      50
1      99
2     130
3    <NA>
dtype: UInt8

## Float data types

Float columns contain numbers with decimal places. The default 'float' size for all platforms is 64 bits (the same size as a C double) and also referred to as 'double-precision'. numpy has additional float sizes of 16 and 32 bits. All float data types can contain missing values. Let's create a Series with floats containing a single missing value and verify the data type.

In [70]:
s_float = pd.Series([4.247, 1234.56789, np.nan])
s_float

0       4.24700
1    1234.56789
2           NaN
dtype: float64

In [39]:
s_float.dtype

dtype('float64')

Below, we change the data type to a 32-bit float, also known as 'single-precision' float. Notice how the second value has changed, as a 32-bit float does not have enough precision to map its value exactly. 

In [71]:
s_float.astype('float32')

0       4.247000
1    1234.567871
2            NaN
dtype: float32

We can use the numpy `finfo` function to get information on each float type. For instance, the 'float32' data type guarantees us 6 significant digits of precision as seen with the `resolution` attribute below.

In [72]:
np.finfo('float32')

finfo(resolution=1e-06, min=-3.4028235e+38, max=3.4028235e+38, dtype=float32)

A 16-bit 'half-precision' float only guarantees 3 digits of precision.

In [73]:
np.finfo('float16')

finfo(resolution=0.001, min=-6.55040e+04, max=6.55040e+04, dtype=float16)

Changing to this data type significantly changes the actual value because of its limited precision.

In [74]:
s_float.astype('float16')

c:\Users\nealv\AppData\Local\Programs\Python\Python311\Lib\site-packages\pandas\io\formats\format.py:1458: RuntimeWarning: overflow encountered in cast
  has_large_values = (abs_vals > 1e6).any()


0       4.246094
1    1235.000000
2            NaN
dtype: float16

## Changing from float to int

So far, we have only changed the size of integer or float data types. We can change types from float to integer and vice-versa. Below, we attempt to go from a 'float64' to an 'int64'. This will fail as the normal integer data type does not allow for missing values.

In [75]:
s_float

0       4.24700
1    1234.56789
2           NaN
dtype: float64

In [45]:
s_float.astype('int64')

IntCastingNaNError: Cannot convert non-finite values (NA or inf) to integer

If you drop the missing values, then conversion is possible. The decimals are truncated and NOT rounded.

In [46]:
s_float.dropna().astype('int64')

0       4
1    1234
dtype: int64

The behavior for the nullable integer data type differs. It does not allow conversion if there any decimal places have values.

In [47]:
s_float.astype('Int64')

TypeError: cannot safely cast non-equivalent object to int64

If you remove the decimals (through rounding for example), then conversion to the nullable integer is possible.

In [48]:
s_float.round(0).astype('Int64')

0       4
1    1235
2    <NA>
dtype: Int64

Going from integer to float will not have as dramatic of an effect and there should be no loss of data as long as the float has enough bits to represent all of the digits.

In [49]:
s_int.astype('float64')

0     50.0
1     99.0
2    130.0
dtype: float64

Conversion from nullable integer to float is also possible. Since floats are numpy data types, they use `NaN` as their missing value representation and not `pd.NA`.

In [50]:
s_nullable_int.astype('float64')

0     50.0
1     99.0
2    130.0
3      NaN
dtype: float64

### Visual display of integers and floats

pandas always display floats with visible decimal points even if there are no significant digits after the decimal. At a minimum, the `.0` will be present. On the other hand, integers are always displayed without a decimal. You can use this rule of thumb to determine the data type without actually accessing the `dtypes` attribute.

## pandas nullable float data type

With the release of pandas 1.2 (December 2020) a nullable float data type was placed in the library. They are referenced by the string `'Float'` (capital F) followed by a bit size of either 16 or 32. Here, we convert the numpy `'float64'` Series to a pandas nullable `'Float64'`.

In [51]:
nullable_float = s_float.astype('Float64')
nullable_float

0         4.247
1    1234.56789
2          <NA>
dtype: Float64

The pandas NA is used for all missing values. Comparison operations again evaluate as a pandas NA.

In [52]:
nullable_float > 100

0    False
1     True
2     <NA>
dtype: boolean

## Boolean data type

Let's now cover the simpler boolean data types. Booleans have a single 8-bit data type in numpy. It makes sense that bit sizes of 16, 32, or 64 don't exist for boolean data types as there are only two possible values, `True` and `False`. You may be curious as to why booleans are not represented with a single bit. This is because a byte (8 bits) is the smallest addressable unit of memory available to modern computers. Let's create a boolean Series using the constructor.

In [53]:
s_bool = pd.Series([True, False])
s_bool

0     True
1    False
dtype: bool

We verify the data type below.

In [54]:
s_bool.dtype

dtype('bool')

### Converting to and from boolean

It is possible to convert integer and float columns to boolean and vice-versa. The only value that will be converted to `False` is 0. All other values are converted to `True`. Use the string 'bool' to convert to boolean. We begin by creating a Series with the integer data type.

In [55]:
s = pd.Series([0, 1, 99, -14])

Call the `astype` method to make the conversion.

In [56]:
s.astype('bool')

0    False
1     True
2     True
3     True
dtype: bool

Series with float data types work the same when converting to boolean. Only the value 0 is converted to `False`. Any other value evaluates as `True`.

In [57]:
s = pd.Series([0, 0.0001, -3.99])
s

0    0.0000
1    0.0001
2   -3.9900
dtype: float64

In [58]:
s.astype('bool')

0    False
1     True
2     True
dtype: bool

Python itself uses the same rules to convert integers and floats to booleans. Let's see a few examples with the built-in `bool` constructor.

In [59]:
bool(5)

True

In [60]:
bool(0)

False

In [61]:
bool(-3.4)

True

In [62]:
bool(0.00000)

False

Converting a boolean Series to integer or float will convert all `True` values to 1 and `False` to 0.

In [63]:
s_bool.astype('int64')

0    1
1    0
dtype: int64

Using a 64-bit integer to store a boolean is overkill. The smallest integer type, `int8` (or `uint8`), can be used to save memory.

In [64]:
s_bool.astype('int8')

0    1
1    0
dtype: int8

## Nullable boolean data type

With the release of pandas 1.0, a new nullable boolean type was made available to support missing values. Until this release, pandas relied on numpy's boolean data type. The nullable boolean is a pandas-only data type, just like the nullable integer. The original boolean data type still exists, but does not support missing values. Let's verify that the original boolean data type cannot contain missing values.

In [65]:
s = pd.Series([True, False, np.nan], dtype='bool')
s

0     True
1    False
2     True
dtype: bool

Surprisingly, this does not fail, but instead converts the numpy `nan` object to `True`. This follows the rule that every non-zero value evaluates as `True` for booleans, even missing values.

Confusingly, constructing a Series of booleans mixed with missing values without explicitly setting the data type preserves the individual type of each value by using the flexible object data type. While this is possible, I advise against mixing data of different types in a single column.

In [66]:
s = pd.Series([True, False, np.nan])
s

0     True
1    False
2      NaN
dtype: object

### Using the new nullable boolean data type

The new nullable boolean data type uses the string 'boolean' to reference it as opposed to 'bool'. Like the nullable integer, it is a pandas-only data type. Here, we construct a Series with the nullable boolean data type.

In [67]:
s = pd.Series([True, False, np.nan], dtype='boolean')
s

0     True
1    False
2     <NA>
dtype: boolean

## Changing data types with an arithmetic operation

Computing an arithmetic operation on a Series can change the resulting Series data type. Division always converts an integer Series to float, even if the result is whole numbers.

In [76]:
s = pd.Series([-15, 45])
s

0   -15
1    45
dtype: int64

The result of using the division operator on an integer Series is always a float.

In [77]:
s / 15

0   -1.0
1    3.0
dtype: float64

This is consistent with core Python which makes the same type conversion.

In [78]:
type(15 / 3)

float

Using floor division, on the other hand, keeps the result as an integer as long as the divisor is an integer.

In [79]:
s // 77

0   -1
1    0
dtype: int64

Multiplying an integer Series by a float always converts it to another float, even if all the resulting values are integers.

In [80]:
s * 4.4

0    -66.0
1    198.0
dtype: float64

Many other combinations of data types paired with operations are possible, but will not be presented here. Instead, it is encouraged for you to test them on your own to view the result.

### Changing data types by assigning new values

As we saw with boolean Series, it is possible to change the data type by assigning one of the values to a different type. Setting one value in an integer Series to a float changes the entire data type of the Series to a float as well.

In [81]:
s

0   -15
1    45
dtype: int64

In [82]:
s.loc[0] = 1.99
s

C:\Users\nealv\AppData\Local\Temp\ipykernel_14024\3992185386.py:1: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '1.99' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.
  s.loc[0] = 1.99


0     1.99
1    45.00
dtype: float64

Setting a non-nullable boolean Series value to the new pandas NA object converts the Series to object and not to the nullable boolean.

In [90]:
s = pd.Series([True, False])
s.loc[0] = pd.NA
s

C:\Users\nealv\AppData\Local\Temp\ipykernel_14024\601854147.py:2: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value 'nan' has dtype incompatible with bool, please explicitly cast to a compatible dtype first.
  s.loc[0] = pd.NA


0      NaN
1    False
dtype: object

From here, you can convert to the nullable boolean with the `astype` method.

In [91]:
s.astype('boolean')

0     <NA>
1    False
dtype: boolean

Converting the data type of an entire Series can be an expensive operation, so it's important to be aware of the possibilities when it happens. There's no need to remember these conversion rules, but to proceed when caution when the possibility arises.

## Setting data types in numpy arrays

We've discussed setting the data type for a pandas Series with the `dtype` parameter. In this section, we will learn how to set the data type for a numpy array, which uses the same parameter `dtype`. You can set it to the string name of the data type you desire. Let's construct an 8-bit integer array. Notice that 150 exceeds the max value of 127.

In [92]:
np.array([1, 5, 150], dtype='int8')

OverflowError: Python integer 150 out of bounds for int8

Here we force numpy to use a 32-bit float. Normally, it would default this array as a 32 or 64 bit integer (depending on your machine).

In [93]:
np.array([1, 5, 150], dtype='float32')

array([  1.,   5., 150.], dtype=float32)

### Creating Series from numpy arrays

The Series constructor accepts one-dimensional numpy arrays as input. The resulting Series will have the same data type as the numpy array.

In [94]:
a = np.array([1, 5, 150], dtype='float32')
pd.Series(a)

0      1.0
1      5.0
2    150.0
dtype: float32

## Different syntax for data types

All of the data type conversions in this chapter were accomplished by using a string such as `'int8'`. There is an alternative approach that can be used. Instead of a string, you can use the actual object itself available directly from numpy or pandas. This section showcases a different method to setting the data type of a Series. For instance, we can use `np.int8` instead of the string `'int8'` to specify a data type.

In [83]:
pd.Series([10, 50]).astype(np.int8)

0    10
1    50
dtype: int8

The following is equivalent to the above.

In [84]:
pd.Series([10, 50]).astype('int8')

0    10
1    50
dtype: int8

All of the numpy data type objects have the same name as their string counterparts. This isn't true for the pandas-only data types. These objects end with the word 'Dtype'. For instance, to convert to a 32-bit nullable integer, you can use `pd.Int32Dtype()`. The parentheses are necessary instantiate this data type.

In [86]:
pd.Series([10, 50]).astype(pd.Int32Dtype()) 

0    10
1    50
dtype: Int32

For the nullable boolean data type, use `pd.BooleanDtype()`.

In [89]:
pd.Series([True, False, np.nan], dtype=pd.BooleanDtype())

0     True
1    False
2     <NA>
dtype: boolean

For nullable 64-bit floats, use `pd.Float64Dtype()`.

In [88]:
pd.Series([7.3, 5.8, np.nan], dtype=pd.Float64Dtype())

0     7.3
1     5.8
2    <NA>
dtype: Float64

### Does it matter which one you use?

You are free to use either one, but I typically use strings when specifying a data type, as numpy must be imported to use the object directly.

## Boolean, integer, and float data type summary

![1]

[1]: images/pandas_numeric_dtypes.png

## Exercises

### Exercise 1

<span style="color:green; font-size:16px">Find the maximum number of a 16-bit integer using arithmetic operations. Then verify it with numpy's `iinfo` function.</span>

In [95]:
np.iinfo('int16')

iinfo(min=-32768, max=32767, dtype=int16)

In [102]:
pd.Series([17000,20000]).astype('int16') + 20000

0   -28536
1   -25536
dtype: int16

In [107]:
 2 ** 15

32768

### Exercise 2

<span style="color:darkgreen; font-size:16px">Construct a Series that has the nullable integer data type. Make sure it has a mix of integers and missing values.</span>

In [103]:
pd.Series([3,6,np.nan])

0    3.0
1    6.0
2    NaN
dtype: float64

In [108]:
pd.Series([3,6,pd.NA])

0       3
1       6
2    <NA>
dtype: object

### Exercise 3

<span style="color:darkgreen; font-size:16px">Take a look at the Series below. Change it's data type such that it uses the least amount of memory and preserves the numbers as they are.</span>

In [109]:
s = pd.Series([1_000, 60_000])

In [112]:
s.info(memory_usage='deep')

<class 'pandas.core.series.Series'>
RangeIndex: 2 entries, 0 to 1
Series name: None
Non-Null Count  Dtype
--------------  -----
2 non-null      int64
dtypes: int64(1)
memory usage: 148.0 bytes


In [117]:
s.astype('uint16').info(memory_usage='deep')

<class 'pandas.core.series.Series'>
RangeIndex: 2 entries, 0 to 1
Series name: None
Non-Null Count  Dtype 
--------------  ----- 
2 non-null      uint16
dtypes: uint16(1)
memory usage: 136.0 bytes


In [116]:
np.iinfo('uint16')

iinfo(min=0, max=65535, dtype=uint16)

### Exercise 4

<span style="color:green; font-size:16px">Find the precision of a 32-bit float and then create a numpy array with values that have decimal places past that precision.</span>

### Exercise 5

<span style="color:green; font-size:16px">Create a Series of numbers that have decimal places. Use the `astype` method to convert it to an integer and then back to a float. Are the decimals from the original Series preserved?</span>

In [118]:
pd.Series([2.2,1.6,23.9]).astype('int16')

0     2
1     1
2    23
dtype: int16

In [119]:
pd.Series([2.2,1.6,23.9]).astype('int16').astype('float16')

c:\Users\nealv\AppData\Local\Programs\Python\Python311\Lib\site-packages\pandas\io\formats\format.py:1458: RuntimeWarning: overflow encountered in cast
  has_large_values = (abs_vals > 1e6).any()


0     2.0
1     1.0
2    23.0
dtype: float16

### Exercise 6

<span style="color:green; font-size:16px">Create a numpy array with 8-bit unsigned integers. Use negative numbers in the construction along with numbers greater than 255. Does the output make sense?</span>

In [125]:
np.array([-50,100,300], dtype='int16')

array([-50, 100, 300], dtype=int16)

### Exercise 7

<span style="color:green; font-size:16px">Create a numpy array that has the values 50 and 100 in it, but do so without actually using those two values (or any operations that create them).</span>

In [126]:
np.array([306,356], dtype='uint8')

OverflowError: Python integer 306 out of bounds for uint8

### Exercise 8

<span style="color:green; font-size:16px">Create a numpy array that contains two integers and the numpy nan missing value. Assign it to a variable name and output it to the screen. What data type is it?</span>

In [127]:
a = np.array([5,10, np.nan]) 

a.dtype

dtype('float64')

### Exercise 9

<span style="color:green; font-size:16px">Construct a Series from the array created in exercise 8. What data type is it? Construct a new Series with the same array forcing it to be a nullable integer.</span>

In [128]:
pd.Series(a)

0     5.0
1    10.0
2     NaN
dtype: float64

In [129]:
pd.Series(a, dtype='Int64')

0       5
1      10
2    <NA>
dtype: Int64

### Exercise 10

<span style="color:green; font-size:16px">Construct a Series of 32-bit nullable integers using the data type object itself (and not the string).</span>

In [131]:
pd.Series([5,10, np.nan], dtype=pd.UInt16Dtype)

C:\Users\nealv\AppData\Local\Temp\ipykernel_14024\652067321.py:1: UserWarning: Instantiating UInt16Dtype without any arguments.Pass a UInt16Dtype instance to silence this warning.
  pd.Series([5,10, np.nan], dtype=pd.UInt16Dtype)


0       5
1      10
2    <NA>
dtype: UInt16